app/controllers/api/connect/v3/systems/products_controller.rb

```diff
@@ -24,19 +24,24 @@ def show
   def migrations
     require_params([:installed_products])
 
-    installed_products = params[:installed_products].map { |hash| product_from_hash(hash) rescue nil }.compact
+    begin
+      upgrade_paths = MigrationEngine.new(@system, installed_products).online_migrations
+
+      render_migration_path upgrade_paths
+    rescue MigrationEngine::MigrationEngineError => e
+      raise ActionController::TranslatedError.new(e.message, *e.data)
+    end
+  end
+
+  def offline_migrations
+    # This part below is not yet working and needs investigation
+    require_params(%i[installed_products target_base_product])
 
     begin
-      upgrade_paths = MigrationEngine.new(@system, installed_products).generate
-
-      render json: (
-        upgrade_paths.reject(&:empty?).map do |item|
-          ActiveModelSerializers::SerializableResource.new(
-            item,
-            each_serializer: ::V3::UpgradePathItemSerializer
-          )
-        end
-      ).to_json
+      offline_upgrade_paths = MigrationEngine.new(@system, installed_products)
+        .offline_migrations(product_from_hash(params[:target_base_product]))
+
+      render_migration_path offline_upgrade_paths
```

Comment:
a bit nick-picky, I would have `render_migration_path` only create the json and keep the `render` in the controller action.

This isn't a blocking comment.